### PR TITLE
Merge RRset operations into master

### PR DESCRIPTION
**Pull request by m0rcq**
operations for RRSets over the wire without full zone unload/load.

* refactored put_zone_rrset/3 -> put_zone_rrset/4
* refactored delete_zone_rrset/3 -> delete_zone_rrset/4
* expanded handling to support zone + FQDN
* abstracted sign_rrset and verify_rrset away from existing _zone()
functions utilising zone structure
* counter table created as "sync_counters"

### CODE REVIEW

File: src/erldns_storage_json.erl [hunk 1]
```diff
@@ -49,6 +49,8 @@ create(Name = host_throttle) ->
 create(Name = lookup_table) ->
   create_ets_table(Name, bag);
 create(Name = handler_registry) ->
+  create_ets_table(Name, set);
```
Comment (aeden) [inline src/erldns_storage_json.erl:52]: You will also need to implement this in `erldns_storage_mnesia`.

File: src/erldns_storage_json.erl [hunk 2]
```diff
@@ -49,6 +49,8 @@ create(Name = host_throttle) ->
...
+create(Name = sync_counters) ->
```
Comment (aeden) [inline src/erldns_storage_json.erl:53]: You will also need to implement this in `erldns_storage_mnesia`.

File: src/erldns_zone_cache.erl [hunk 3]
```diff
@@ -223,9 +263,59 @@ put_zone(Name, Zone) ->
...
+	  true -> 
+		  lager:debug("Processing RRSet (~p) for Zone (~p): counter (~p) provided is higher than system", [RRFqdn, ZoneName, Counter]),
+		  % TODO: add custom types lookup
+		  Type = erldns_records:name_type(RRSetType),
+		  case find_zone_in_cache(erldns:normalize_name(ZoneName)) of
+			{ok, Zone} -> Zone,
+			  lager:debug("Putting RRSet (~p) with Type: ~p for Zone (~p): ~p", [RRFqdn, Type, ZoneName, Records]),
+			  lager:debug("Available Keys: ~p", [Keys]),
+			  % if keys are missing from RRset payload, use existing zone keysets
+			  KeySets = case length(Keys) of
```
Comment (aeden) [inline src/erldns_zone_cache.erl:287]: For simplicity always use the keys from the zone for now.

File: src/erldns_zone_cache.erl [hunk 4]
```diff
@@ -362,6 +463,34 @@ verify_zone(_Zone, DnskeyRRs, KeyRRSigRecords) ->
...
+  lager:debug("KeyRRSigRecords: ~p", [KeyRRSigRecords]),
+  verify_rrset(DnsKeyRRs, KeyRRSigRecords),
+  %RRSetRecords ++ KeyRRSigRecords ++ rewrite_soa_rrsig_ttl(Records, KeyRRSigRecords).
+  RRSetRecords.
+	
+% Verify RRSet
+-spec(verify_rrset([dns:rr()], [dns:rr()]) -> boolean()).
+verify_rrset(DnsKeyRRs, KeyRRSigRecords) ->
+  lager:debug("Verify RRSet"),
+  case lists:filter(fun(RR) -> RR#dns_rr.data#dns_rrdata_dnskey.flags =:= 257 end, DnsKeyRRs) of
```
Comment (aeden) [inline src/erldns_zone_cache.erl:482]: `include/erldns.hrl` includes a macro for `DNSKEY_KSK_TYPE` that you can use here.

File: src/erldns_storage_json.erl (hunk 2 above)
Comment (m0rcq) [reply src/erldns_storage_json.erl:53]: commit: https://github.com/dnsimple/erldns/commit/4a9219b8b55c2a50dd74bbb7dc264f306497ceec

covers mnesia storage backend.

File: src/erldns_zone_cache.erl (hunk 3 above)
Comment (m0rcq) [reply src/erldns_zone_cache.erl]: commit: https://github.com/dnsimple/erldns/commit/fff79c3ec3cabb1b1783d528533ee0e7f704942d 

covers this (use existing Zone keysets only)

File: src/erldns_zone_cache.erl (hunk 4 above)
Comment (m0rcq) [reply src/erldns_zone_cache.erl]: commit: https://github.com/dnsimple/erldns/commit/ac0506cd6d09897ae14f38b8bdaa474c421abb58 addresses this.

**Comment by m0rcq** [comment]
Suggestions are welcome to fix the spec below trying to capture two different argument patterns.

https://github.com/dnsimple/erldns/pull/100/commits/3b0c4036e5cbc1c161fe5997df0d13c474e9f846#diff-8a8e060b9a8de98f1aa024db2afad89bR266

**Comment by m0rcq** [comment]
@aeden:

map_rrset_type/1 has been moved to erldnsimple_zone_parser.erl and I am trying to find a clean way to inject the function into erldns_zoneclient on startup, but looking into the current flow:

https://github.com/dnsimple/erldnsimple/blob/master/src/erldnsimple.erl#L12
[...]
  application:ensure_all_started(erldnsimple).

the erldns_zoneclient:start() is not called explicitly, but relies on the application specification defined in 
https://github.com/dnsimple/erldnsimple/blob/master/src/erldnsimple.app.src#L16.

as per specification: http://erlang.org/doc/man/app.html

The mod accepts StartArgs but only for modules in the current application i.e erldnsimple, where erldns_zoneclient application fails out of this scope.

the start_phases can be used as it will call the primary application, plus all the dependent applications, but they will need to have Module:start_phases/3 defined which is not the right path for such a simple feat as providing higher-order fun to another application.

The erldns_zoneclient_app creates the supervisor in start/2
https://github.com/dnsimple/erldns-zoneclient/blob/master/src/erldns_zoneclient_app.erl#L25 

and the supervisor creates the erldns_zoneclient_server_monitor child process here:
https://github.com/dnsimple/erldns-zoneclient/blob/master/src/erldns_zoneclient_sup.erl#L47

On top of this we have the actual function which handles websocket ingress defined as higher-order fun for the websocket_client library.

https://github.com/dnsimple/erldns-zoneclient/blob/master/src/erldns_zoneclient_zoneserver_monitor.erl#L137

As we have multiple logic points mixing application + supervisors behaviour, what do you think will be cleanest approach to inject a function from the parent process (erldnsimple)?

Should I try to:

* have explicit erldns_zoneclient:start(Args) in erldnsimple:start() instead of relying on the application spec

* explicit init(Args) in erldnsimple_sup Supervisor:
https://github.com/dnsimple/erldnsimple/blob/master/src/erldnsimple_sup.erl#L19

Or some other approach?

**Comment by aeden** [comment]
I think you'll likely need an approach similar to how handlers are injected into erldns: https://github.com/dnsimple/erldns/blob/master/src/erldns_handler.erl

**Comment by m0rcq** [comment]
> I think you'll likely need an approach similar to how handlers are injected into erldns: https://github.com/dnsimple/erldns/blob/master/src/erldns_handler.erl

Roger that - will implement using similar approach @ gen_server's erldns_zoneclient_zoneserver_monitor

**Comment by m0rcq** [comment]
> I think you'll likely need an approach similar to how handlers are injected into erldns: https://github.com/dnsimple/erldns/blob/master/src/erldns_handler.erl

@aeden - relevant PRs:

https://github.com/dnsimple/erldnsimple/pull/113
https://github.com/dnsimple/erldns-zoneclient/pull/5

I did explicit init call on app startup to erldnsimple_zone_parser:init/1 which registers the mapping function map_rrset_type/1 to erldns_zoneclient_zoneserver_monitor's gen_server.

Is the _zoneserver_monitor the ideal place to register mappers? - I think not, but we don't have any better candidates with gen_server behaviour @ zoneclient (zonechecker/fetcher are not any better candidates). At least _monitor is driving the websocket related functionality.

websocket_handler/3 is calling erldns_zoneclient_zoneserver_monitor:map_rrset_type/1 to do the required data transformation.

websocket_handler/3 has been refactored slightly to guard against bogus data coming via the wire.

I managed to crash the process handling websocket ingress in the scenario of sending incomplete JSON payload and the supervisor did not spawn it back even after timeout, so I think we will need to look into this and understand why.

websocket_client:start_link(WebsocketUrl, erldns_zoneclient_websocket_handler, [])

File: src/erldns_zone_cache.erl [hunk 5]
```diff
@@ -6,7 +6,7 @@
 %%
 %% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 %% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
-%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
```
Comment (aeden) [inline src/erldns_zone_cache.erl:9]: Accidental removal?

File: src/erldns_zone_cache.erl [hunk 6]
```diff
@@ -23,6 +23,7 @@
 
 -include_lib("dns_erlang/include/dns.hrl").
 -include("erldns.hrl").
+-include_lib("stdlib/include/ms_transform.hrl").
```
Comment (aeden) [inline src/erldns_zone_cache.erl:26]: Still used?

File: src/erldns_zone_cache.erl [hunk 7]
```diff
@@ -249,7 +346,73 @@ delete_zone_records(Name) ->
...
+-spec delete_zone_rrset(binary(), binary(), binary(), integer()) -> any().
+delete_zone_rrset(Name, Digest, RRFqdn, Type) ->
+  delete_zone_rrset(Name, Digest, RRFqdn, Type, 0).
+
+%% @doc Remove zone RRSet
+-spec delete_zone_rrset(binary(), binary(), binary(), integer(), integer()) -> any().
+delete_zone_rrset(ZoneName, Digest, RRFqdn, Type, Counter) ->
+  case find_zone_in_cache(erldns:normalize_name(ZoneName)) of
+    {ok, Zone} -> 
+	  Zone,
```
Comment (aeden) [inline src/erldns_zone_cache.erl:359]: You shouldn't need this as the `Zone` is matched above and thus available in the block.

File: src/erldns_zone_cache.erl [hunk 8]
```diff
@@ -249,7 +346,73 @@ delete_zone_records(Name) ->
...
+	  CurrentCounter = get_rrset_sync_counter(ZoneName, RRFqdn, Type),
+	  if
```
Comment (aeden) [inline src/erldns_zone_cache.erl:361]: Stylistically I avoid `if` in favor of case matching whenever possible.

File: src/erldns_zone_cache.erl [hunk 9]
```diff
@@ -249,7 +346,73 @@ delete_zone_records(Name) ->
...
+		  RRSigRecs = filter_rrsig_records_with_type_covered(RRFqdn, Type),
+		  % delete all RRSIGs for FQDN 
+		  erldns_storage:select_delete(zone_records_typed, [{{{erldns:normalize_name(ZoneName), erldns:normalize_name(RRFqdn), ?DNS_TYPE_RRSIG_NUMBER}, '_'},[],[true]}]),
+		  % write back the filtered RRSIG set
+		  lists:map(fun(R) ->
+			erldns_storage:insert(zone_records_typed, R) end,
+			RRSigRecs),
+		  % only write counter if called explicitly with Counter value i.e. different than 0.
+		  % this will not write the counter if called by put_zone_rrset/3 as it will prevent subsequent delete ops
+		  if
```
Comment (aeden) [inline src/erldns_zone_cache.erl:377]: As above, I generally use case matching whenever possible rather than `if`.

**Review by aeden — APPROVED** [review]
A few cleanup items for you, feel free to apply as you see fit.

File: src/erldns_zone_cache.erl (hunk 5 above)
Comment (m0rcq) [reply src/erldns_zone_cache.erl:9]: most likely - fixing.

File: src/erldns_zone_cache.erl (hunk 6 above)
Comment (m0rcq) [reply src/erldns_zone_cache.erl:26]: good catch - the ets:fun2ms/1's calls have been removed.

File: src/erldns_zone_cache.erl (hunk 7 above)
Comment (m0rcq) [reply src/erldns_zone_cache.erl:359]: the zone is matched in put_zone_rrset/4 and we can pass it to delete_zone_rrset/4 within that function, but this is not the case when we call delete_zone_rrset/5 on RRset's delete operation from zoneclient, hence the explicit call within the function.

File: src/erldns_zone_cache.erl (hunk 9 above)
Comment (m0rcq) [reply src/erldns_zone_cache.erl:377]: Agreed, I will refactor the Ifs.

Comment (m0rcq) [reply src/erldns_zone_cache.erl]: commit: https://github.com/dnsimple/erldns/pull/100/commits/1788d173d41c09fc5ab6780a5d029d62c5d2cd94 refactors the ifs to case of expressions

**Pull request merged**